### ISSUE TITLE
[question] How to make transitive dependencies work properly

### ISSUE THREAD

**Issue by dkumsh** [conan-io/conan#11532]
Hi 
I am using conan 1.49 (latest at the moment).

I have project called **base**, which uses number of third party libraries. 
It uses cmake to produce two libraries: **libbase.a** and **libbase-shared.so**.
I am using _CMakeToolchain_ generator for seamless _CLion_ integration and 
_CMakeDeps_ generator to produce  dependencies modules for cmake:

```python
from conan import ConanFile
from conan.tools.cmake import CMakeToolchain, CMakeDeps, CMake, cmake_layout
    requires = ( "boost/1.73.0",  ... "spdlog/1.9.0", "gtest/1.10.0"  )
    def generate(self):
        CMakeToolchain(self, generator="Ninja").generate()
        CMakeDeps(self).generate()

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()

    def package(self):
        CMake(self).install()
```

For consumers of the project **base** I define:
```python
    def package_info(self):
        req = [ "boost::boost", ... , "spdlog::spdlog", "gtest::gtest"]
        self.cpp_info.includedirs = ["include/base"]
        #
        self.cpp_info.components["static"].set_property("cmake_target_name","APP::base-static")
        self.cpp_info.components["static"].libdirs = ["lib"]
        self.cpp_info.components["static"].includedirs = ["include/base"]
        self.cpp_info.components["static"].libs = ["base-static"]
        self.cpp_info.components["static"].requires = req
        #
        self.cpp_info.components["base-shared"].set_property("cmake_target_name","APP::base-shared")
        self.cpp_info.components["shared"].libdirs = ["lib"]
        self.cpp_info.components["shared"].libs = ["base"]
        self.cpp_info.components["shared"].includedirs = ["include/base"]
        self.cpp_info.components["shared"].requires = req
```
Now I added a second project (let us call it **consumer1**). It depends on **base**. 
It is very similar to **base** project. It also produces a static and a shared library. 
It builds fine so far. 

The main difference in **consumer1**'s recipe is that it requires **base** (and therefore indirectly depends on dependencies of **base**) and it also depends on an extra library _extralib_:
```python
    requires = ( "base/1.0.0", "extralib/1.2.3" )
```
 **Consumer1** has the following **package_info**:
```python
    def package_info(self):
        self.cpp_info.libdirs = ["lib"]
        self.cpp_info.includedirs = ["include/consumer1"]
        #
        self.cpp_info.components["shared"].set_property("cmake_target_name","APP::consumer1-shared")
        self.cpp_info.components["shared"].libdirs = ["lib"]
        self.cpp_info.components["shared"].libs = ["consumer1"]
        self.cpp_info.components["shared"].includedirs = ["include/consumer1"]
        self.cpp_info.components["shared"].requires = [ what should be here? ]
        #
        self.cpp_info.components["static"].set_property("cmake_target_name","APP::consumer1-static")
        self.cpp_info.components["static"].libdirs = ["lib"]
        self.cpp_info.components["static"].libs = ["consumer1-static"]
        self.cpp_info.components["static"].includedirs = ["include/consumer1"]
        #
        # this is what I need  but 'conan create' fails in consumer2 (consumer of this project):
        # self.cpp_info.components["static"].requires = [ "APP::base-static", "extralib::extralib" ]
        #
        #  if I define it the following way, this fixes "conan create"  failure 
        #  however it fails to link a this dependency 
        # as it transitively depends on both libbase.a (static) and libbase-shared.so  libraries 
        # self.cpp_info.components["shared"].requires = [ "base::base", "extralib::extralib" ]
        #
        self.cpp_info.components["static"].requires = [ what should be here? ]
```
If then I define a **consumer2** project consuming **consumer1** 

```python
    requires = ( "consumer1/1.0.0" )
```

If say I define in **consumer2** an executable **foo** statically linking to **libconsumer1.a** I should be able to define target _APP::consumer1-static_ for this static library and then in **consumer1** project I need to be able to specify that component _APP::consumer1-static_ depends on component _APP::base-static_. Similarly _APP::consumer1-shared_ should depend on _APP::base-shared_. **How would I do that correctly?**

In my scenario if I define in **consumer1** as follows ( see the comments above)
```python
      self.cpp_info.components["static"].requires = [ "APP::base-static", "extralib::extralib" ]
```
I will get the following error when building **consumer2**:
```log
consumer2/1.0.0@user/channel: ERROR: Traceback (most recent call last):
  File "/root/workspace/consumer2/venv/lib/python3.8/site-packages/conan/tools/cmake/cmakedeps/templates/__init__.py", line 43, in render
    context = self.context
  File "/root/workspace/consumer2/venv/lib/python3.8/site-packages/conan/tools/cmake/cmakedeps/templates/target_data.py", line 32, in context
    components = self._get_required_components_cpp()
  File "/root/workspace/consumer2/venv/lib/python3.8/site-packages/conan/tools/cmake/cmakedeps/templates/target_data.py", line 153, in _get_required_components_cpp
    req = direct_visible_host[pkg]
  File "/root/workspace/consumer2/venv/lib/python3.8/site-packages/conans/model/dependencies.py", line 76, in __getitem__
    return self._data[r]
KeyError: {'ref': APP/unknown@unknown/unknown, 'build': False, 'direct': True, 'test': False, 'visible': True}

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/workspace/consumer2/venv/lib/python3.8/site-packages/conans/client/generators/__init__.py", line 177, in write_generators
    generator.generate()
  File "/root/workspace/consumer2/venv/lib/python3.8/site-packages/conan/tools/cmake/cmakedeps/cmakedeps.py", line 46, in generate
    generator_files = self.content
  File "/root/workspace/consumer2/venv/lib/python3.8/site-packages/conan/tools/cmake/cmakedeps/cmakedeps.py", line 88, in content
    self._generate_files(require, dep, ret, find_module_mode=False)
  File "/root/workspace/consumer2/venv/lib/python3.8/site-packages/conan/tools/cmake/cmakedeps/cmakedeps.py", line 101, in _generate_files
    ret[data_target.filename] = data_target.render()
  File "/root/workspace/consumer2/venv/lib/python3.8/site-packages/conan/tools/cmake/cmakedeps/templates/__init__.py", line 45, in render
    raise ConanException("error generating context for '{}': {}".format(self.conanfile, e))
conans.errors.ConanException: error generating context for 'consumer1/1.0.0@user/channel': {'ref': APP/unknown@unknown/unknown, 'build': False, 'direct': True, 'test': False, 'visible': True}

ERROR: Error in generator 'CMakeDeps': error generating context for 'consumer1/1.0.0@user/channel': {'ref': APP/unknown@unknown/unknown, 'build': False, 'direct': True, 'test': False, 'visible': True}

```
 
 I wonder if this is a bug or am I doing something  wrong?
 
- [x ] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @dkumsh 

I would like first to understand better your use case, can you please clarify:

- Are your "components" "static" and "shared" mutually exclusive? Are they 2 different builds of the same source code? Can you link an executable from downstream to both simultaneously?
- If the answer is that they are exclusive, and you can only link 1 of them, it seems that is not the case for ``components`` and you might be abusing them
- Why wouldn't you use the ``shared`` option to represent different static/shared builds?

Thanks

**Comment by dkumsh** [comment]
Hi @memsharded 
For **base** project "static" and "shared" are built from almost the same cpp files.  
Specifically, shared version of the library has everything that static library has plus extra code defining python bindings for the common part of both libs.  
Historically (before conan use) these two flavours of the libraries (static and shared) are built together in one single build (both are defined in the same CMakeList.txt and  built by the same cmake build command). 
However on consumer side each target normally needs either shared or static versions of the libraries, 
i.e. potentially they could be made exclusive via use of `shared` option. 
This will require an extra effort to migrate though. 
A minor advantage of combined build is also that we do not need to compile twice the cpp files common to both static and shared versions of the library ( which is majority of the files).  
Potentially I could consider introducing `shared` option,  but here it is an attempt to take `components` route. 
I do see the advantages of switching to `shared` options
I do not see however why the `components` approach should not work here too.

**Comment by memsharded** [comment]
> For base project "static" and "shared" are built from almost the same cpp files.
Specifically, shared version of the library has everything that static library has plus extra code defining python bindings for the common part of both libs

It is fine. The model of Conan binary packages is that they share the same source code and conanfile.py recipe, and just have different variants of configurations. But they can be installed and managed separately

> Historically (before conan use) these two flavours of the libraries (static and shared) are built together in one single build (both are defined in the same CMakeList.txt and built by the same cmake build command).

If the compile time is the concern, there is a feature, the ``build_id()``, that will allow building only once, and packaging 2 different Conan package binaries, one for each configuration (shared and static)

> A minor advantage of combined build is also that we do not need to compile twice the cpp files common to both static and shared versions of the library ( which is majority of the files).

See the above about ``build_id()``

> However on consumer side each target normally needs either shared or static versions of the libraries,

> I do not see however why the components approach should not work here too.

Some reasons:
- Components are not designed to be exclusive or binary variants, but different components of a group, like "libsparse" and "libdense" of a "math" package.
- One of the core features of Conan is the management of binary variants. All the package_id logic is based on the management of those variants.
- There is logic in Conan that works differently if the ``shared`` option exists. There will be even more in Conan 2.0 (see https://www.youtube.com/watch?v=kKGglzm5ous)
- Components will be collapsed (assuming they are not mutually exclusive, but complementary) in many generators that don't support components yet (MSBuildDeps, for example), meaning that this would be a broken package for other generators
- Other features like "collect_libs", and probably some of the ``package_info()`` features are also broken for this assumption. This might be the issues that you are experiencing.

Overall, taking into account that you don't really need to break your build and you can use the ``build_id()`` method, I'd say that it is going to take less effort and less future maintenance to have one package per shared/static configuration with the ``shared`` option.

**Comment by dkumsh** [comment]
Thanks for you reply and for sharing the video. Will watch it. 
This all make sense in my case and I am ready to try this approach. 
I still have some questions left:

- base library depends on many other third party libraries. Most of them must be linked statically even with my "shared" library.  How can I control `shared` option of the dependencies independently of the `shared` option of the consumer project?
- in IDE (CLion) I would need to build exclusively static or shared library. How would I manage that? 
  It's even more complicated as I have chain of at layers of such libraries (each layer is consumed by the 
  next one: base <- consumer1 <- consumer2 ...). I.e. say to debug a statically linked app from consumer2 I would need to 
  install static base and consumer1 and to debug a shared library I need to re-install them for CLion. 
  In other words even  even though these static/shared packages seem to be "exclusive" from usage perspective, 
  from development/CLIon perspective it does not seem convenient.
 - regardless of the conversation here about "component" vs. "option" approaches, there might be mode "valid" cases for components. I wanted to understand if the issue with transitive use of components - did I do something incorrectly or it's just a bug?  
 
 Many thanks

**Comment by memsharded** [comment]
Hi @dkumsh 

> base library depends on many other third party libraries. Most of them must be linked statically even with my "shared" library. How can I control shared option of the dependencies independently of the shared option of the consumer project?

all options are per-package. You can specify them using patterns, both in the profiles/command line or in recipes (possible, but maybe better consider profile/command line first), as ``pkg1*:shared=True``, ``pkg2*:shared=False``, ``pkg3*:shared=False``, etc.

> in IDE (CLion) I would need to build exclusively static or shared library. How would I manage that?
It's even more complicated as I have chain of at layers of such libraries (each layer is consumed by the
next one: base <- consumer1 <- consumer2 ...). I.e. say to debug a statically linked app from consumer2 I would need to
install static base and consumer1 and to debug a shared library I need to re-install them for CLion.
In other words even even though these static/shared packages seem to be "exclusive" from usage perspective,
from development/CLIon perspective it does not seem convenient.

This is more a build system thing than an IDE thing. I assume you are using CLion with CMake, and it also depends about the platform (if multi-config as VS or single-config as Makefiles). I'd strongly recommend to have a look to the latest Conan ``CMakeToolchain`` generator (together with ``CMakeDeps``) in Conan 1.49, as it can generate multiple ``CMakePresets`` files, that can also be used by CLion. This approach allows to install as as many different configurations (Debug/Release, static/shared, different compilers or architectures) in a consistent way, and manage them. You need latest versions of CMake, Conan and CLion, but this new integration is looking fantastic so far. Have a look at: https://docs.conan.io/en/latest/reference/conanfile/tools/cmake.html, and more specifically at: https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmaketoolchain.html#using-the-toolchain-in-developer-flow and https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmake_layout.html

> regardless of the conversation here about "component" vs. "option" approaches, there might be mode "valid" cases for components. I wanted to understand if the issue with transitive use of components - did I do something incorrectly or it's just a bug?

I think the mistake there could be mixing the CMake target name ``APP::consumer1-shared`` with the actual component name. When you are using the ``components.requires``, it should be pointing to components names, not to CMake target names.

**Comment by dkumsh** [comment]
Thanks for detailed answers and review of the approach I used. 
Indeed using actual component name ( in my case `base::static` instead of `APP::base-static`) fixes the problem with transitive dependencies.
However based on our conversation I like the approach with building mutually exclusive shared or static libs.  
I  am going to switch to this approach.
Thanks again

**Comment by memsharded** [comment]
Thanks to you for all the feedback.
Don't hesitate to open new tickets if you need further help or have more questions.
